Firestore/core/test/firebase/firestore/local/mutation_queue_test.cc

```diff
@@ -0,0 +1,432 @@
+    ASSERT_EQ(2, BatchCount());
+
+    mutation_queue_->RemoveMutationBatch(batch1);
+    ASSERT_EQ(1, BatchCount());
+
+    mutation_queue_->RemoveMutationBatch(batch2);
+    ASSERT_EQ(0, BatchCount());
+    ASSERT_TRUE(mutation_queue_->IsEmpty());
+  });
+}
+
+TEST_P(MutationQueueTest, AcknowledgeBatchID) {
+  persistence_->Run("test_acknowledge_batch_id", [&] {
+    ASSERT_EQ(BatchCount(), 0);
+
+    MutationBatch batch1 = AddMutationBatch();
+    MutationBatch batch2 = AddMutationBatch();
+    MutationBatch batch3 = AddMutationBatch();
+    ASSERT_GT(batch1.batch_id(), kBatchIdUnknown);
+    ASSERT_GT(batch2.batch_id(), batch1.batch_id());
+    ASSERT_GT(batch3.batch_id(), batch2.batch_id());
+
+    ASSERT_EQ(BatchCount(), 3);
+
+    mutation_queue_->AcknowledgeBatch(batch1, {});
+    mutation_queue_->RemoveMutationBatch(batch1);
+    ASSERT_EQ(BatchCount(), 2);
+
+    mutation_queue_->AcknowledgeBatch(batch2, {});
+    ASSERT_EQ(BatchCount(), 2);
+
+    mutation_queue_->RemoveMutationBatch(batch2);
+    ASSERT_EQ(BatchCount(), 1);
+
+    mutation_queue_->RemoveMutationBatch(batch3);
+    ASSERT_EQ(BatchCount(), 0);
+  });
+}
+
+TEST_P(MutationQueueTest, AcknowledgeThenRemove) {
+  persistence_->Run("test_acknowledge_then_remove", [&] {
+    MutationBatch batch1 = AddMutationBatch();
+
+    mutation_queue_->AcknowledgeBatch(batch1, {});
+    mutation_queue_->RemoveMutationBatch(batch1);
+
+    ASSERT_EQ(BatchCount(), 0);
+  });
+}
+
+TEST_P(MutationQueueTest, LookupMutationBatch) {
+  // Searching on an empty queue should not find a non-existent batch
+  persistence_->Run("test_lookup_mutation_batch", [&] {
+    absl::optional<MutationBatch> not_found =
+        mutation_queue_->LookupMutationBatch(42);
+    ASSERT_EQ(not_found, absl::nullopt);
+
+    std::vector<MutationBatch> batches = CreateBatches(10);
+    std::vector<MutationBatch> removed = RemoveFirstBatches(3, &batches);
+
+    // After removing, a batch should not be found
+    for (size_t i = 0; i < removed.size(); i++) {
+      not_found = mutation_queue_->LookupMutationBatch(removed[i].batch_id());
+      ASSERT_EQ(not_found, absl::nullopt);
+    }
+
+    // Remaining entries should still be found
+    for (const MutationBatch& batch : batches) {
+      absl::optional<MutationBatch> found =
+          mutation_queue_->LookupMutationBatch(batch.batch_id());
+      ASSERT_EQ(found->batch_id(), batch.batch_id());
+    }
+
+    // Even on a nonempty queue searching should not find a non-existent batch
+    not_found = mutation_queue_->LookupMutationBatch(42);
+    ASSERT_EQ(not_found, absl::nullopt);
+  });
+}
+
+TEST_P(MutationQueueTest, NextMutationBatchAfterBatchID) {
+  persistence_->Run("test_next_mutation_batch_after_batch_id", [&] {
+    std::vector<MutationBatch> batches = CreateBatches(10);
+    std::vector<MutationBatch> removed = RemoveFirstBatches(3, &batches);
+
+    for (size_t i = 0; i < batches.size() - 1; i++) {
+      const MutationBatch& current = batches[i];
```

Comment:
Very optional: it _seems_ like it would be very slightly less verbose with iterators:
```cpp
  for (auto current = batches.begin(); current != batches.end() - 1; ++current) {
      auto next = current + 1;
      absl::optional<MutationBatch> found =
          mutation_queue_->NextMutationBatchAfterBatchId(current->batch_id());
      ASSERT_EQ(found->batch_id(), next->batch_id());
    }
```

Applies to some of the following loops as well.

Reply:
I opted not to do this. Dissimilarity to the other ports outweighs the slight gain.